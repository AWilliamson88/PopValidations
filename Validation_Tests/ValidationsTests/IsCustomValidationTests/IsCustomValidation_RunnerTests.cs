using ApprovalTests;
using FluentAssertions;
using System.Threading.Tasks;
using PopValidations;
using PopValidations_Tests.TestHelpers;
using PopValidations_Tests.ValidationsTests.GenericTestableObjects;
using Xunit;

namespace PopValidations_Tests.ValidationsTests.IsCustomValidationTests;

public class IsCustom_AllErrored_TestingValidator : AbstractValidator<NullAllFieldTypesDto>
{
    public IsCustom_AllErrored_TestingValidator()
    {
        Describe(x => x.Integer).Is("Description", "Error", IntTest);
        Describe(x => x.String).Is("Description", "Error", StringTest);
        Describe(x => x.Decimal).Is("Description", "Error", DecimalTest);
    }

    public static bool IntTest(int? value)
    {
        return value == 1;
    }

    public static bool StringTest(string? value)
    {
        return value == "1";
    }

    public static bool DecimalTest(decimal? value)
    {
        return value == 1;
    }
}

public class IsCustomValidation_RunnerTests
{
    [Fact]
    public async Task GivenAValidator_WithNoErrors_ThenNoErrorsAreProduced()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(
            new IsCustom_AllErrored_TestingValidator()
        );

        // Act
        var validationResult = await runner.Validate(
            new NullAllFieldTypesDto()
            {
                Integer = 1,
                String = "1",
                Decimal = 1
            }
        );

        // Assert
        validationResult.Errors.Should().BeEmpty();
    }

    [Fact]
    public async Task GivenAValidator_WithErrors_ThenEveryFieldHasErrors()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(
            new IsCustom_AllErrored_TestingValidator()
        );

        // Act
        var validationResult = await runner.Validate(
            new NullAllFieldTypesDto()
            {
                Integer = 2,
                String = "2",
                Decimal = 2
            }
        );
        var json = JsonConverter.ToJson(validationResult);

        // Assert
        validationResult.Errors.Should().HaveCount(3);
        Approvals.VerifyJson(json);
    }

    [Fact]
    public void GivenAValidator_WhenDescribing_ThenEveryFieldHasDescriptions()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(
            new IsCustom_AllErrored_TestingValidator()
        );

        // Act
        var description = runner.Describe();
        var json = JsonConverter.ToJson(description);

        // Assert
        description.Results.Should().HaveCount(3);
        Approvals.VerifyJson(json);
    }
}
