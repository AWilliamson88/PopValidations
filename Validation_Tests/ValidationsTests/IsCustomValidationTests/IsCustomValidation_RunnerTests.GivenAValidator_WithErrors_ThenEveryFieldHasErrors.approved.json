{
    "errors": {
        "integer": [
            "Error"
        ],
        "string": [
            "Error"
        ],
        "decimal": [
            "Error"
        ]
    }
}